zerver/views/streams.py

```diff
@@ -648,6 +650,164 @@ def you_were_just_subscribed_message(
 RETENTION_DEFAULT: str | int = "realm_default"
 
 
+@transaction.atomic(savepoint=False)
+@require_non_guest_user
+@typed_endpoint
+def add_channel(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    *,
+    announce: Json[bool] = False,
+    can_add_subscribers_group: Json[int | UserGroupMembersData] | None = None,
+    can_administer_channel_group: Json[int | UserGroupMembersData] | None = None,
+    can_move_messages_out_of_channel_group: Json[int | UserGroupMembersData] | None = None,
+    can_move_messages_within_channel_group: Json[int | UserGroupMembersData] | None = None,
+    can_remove_subscribers_group: Json[int | UserGroupMembersData] | None = None,
+    can_resolve_topics_group: Json[int | UserGroupMembersData] | None = None,
+    can_send_message_group: Json[int | UserGroupMembersData] | None = None,
+    can_subscribe_group: Json[int | UserGroupMembersData] | None = None,
+    channel: Annotated[Json[AddSubscriptionData], ApiParamConfig("channel")],
+    folder_id: Json[int] | None = None,
+    history_public_to_subscribers: Json[bool] | None = None,
+    invite_only: Json[bool] = False,
+    is_web_public: Json[bool] = False,
+    is_default_stream: Json[bool] = False,
+    message_retention_days: Json[str] | Json[int] = RETENTION_DEFAULT,
+    send_new_subscription_messages: Json[bool] = True,
+    subscribers: Json[list[str] | list[int]],
+    topics_policy: Json[
+        Annotated[
+            str | None,
+            AfterValidator(
+                lambda val: parse_enum_from_string_value(
+                    val,
+                    "topics_policy",
+                    StreamTopicsPolicyEnum,
+                )
+            ),
+        ]
+    ] = None,
+) -> HttpResponse:
+    realm = user_profile.realm
+    ensure_stream_does_not_exist_already(channel.name, realm)
```

Comment:
This doesn't call `check_stream_name` and thus would I think be a security bug.

I think you can just use `check_stream_name_available`; I don't understand why you're duplicating that code. `check_stream_name_available` could raise `ChannelExistsError` rather than a generic `JsonableError` easily.